pallets/foreign-investments/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+		/// `TokenSwapOrderId`.
+		InvestmentInfoNotFound,
+		/// Failed to retrieve the `TokenSwapReason` from the given
+		/// `TokenSwapOrderId`.
+		TokenSwapReasonNotFound,
+		/// Failed to transition the `InvestState`.
+		InvestError(InvestError),
+		/// Failed to transition the `RedeemState.`
+		RedeemError(RedeemError),
+	}
+
+	impl<T> From<InvestError> for Error<T> {
+		fn from(error: InvestError) -> Self {
+			Error::<T>::InvestError(error)
+		}
+	}
+
+	impl<T> From<RedeemError> for Error<T> {
+		fn from(error: RedeemError) -> Self {
+			Error::<T>::RedeemError(error)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Attempts to transition an `InvestState` after an epoch execution:
+		/// * If the state includes `InvestmentOngoing` and the unprocessed
+		///   investment amount is zero, removes `InvestmentOngoing`.
+		/// * If the state includes `InvestmentOngoing` and the unprocessed
+		///   investment amount is positive, updates the `invest_amount` to the
+		///   unprocessed one.
+		///
+		/// NOOP: If the unprocessed investment amount is zero or the state does
+		/// not include `InvestmentOngoing`
+		#[pallet::weight(100_000_000 + T::DbWeight::get().reads_writes(5, 5).ref_time())]
+		#[pallet::call_index(1)]
+		pub fn nudge_invest_state(
+			origin: OriginFor<T>,
+			investor: T::AccountId,
+			investment_id: T::InvestmentId,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			if let Some(invest_state) = InvestmentState::<T>::get(&investor, investment_id) {
+				let amount_unprocessed_investment =
+					T::Investment::investment(&investor, investment_id)?;
+				let new_state = invest_state
+					.transition(types::InvestTransition::EpochExecution(
+						amount_unprocessed_investment,
+					))
+					.map_err(|e| {
+						log::debug!("InvestState transition error: {:?}", e);
+						Error::<T>::from(InvestError::EpochExecution)
+					})?;
+				Pallet::<T>::apply_invest_state_transition(&investor, investment_id, new_state)?;
+			}
+			Ok(())
+		}
+
+		/// Attempts to transition a `RedeemState` after an epoch execution:
+		/// * If the inner state includes `Redeeming` and the unprocessed
+		///   redemption amount is zero, removes `Redeeming`.
+		/// * If the inner state includes `Redeeming` and the unprocessed
+		///   redemption amount is positive, updates the amount to the
+		///   unprocessed one.
+		///
+		/// NOOP: If the unprocessed redemption amount is zero or the inner
+		/// state does not include `Redeeming`.
+		#[pallet::weight(100_000_000 + T::DbWeight::get().reads_writes(5, 5).ref_time())]
+		#[pallet::call_index(2)]
+		pub fn nudge_redeem_state(
+			origin: OriginFor<T>,
+			investor: T::AccountId,
+			investment_id: T::InvestmentId,
+		) -> DispatchResult {
+			ensure_signed(origin)?;
+
+			if let Some(redeem_state) = RedemptionState::<T>::get(&investor, investment_id) {
+				let amount_unprocessed_redemption =
+					T::Investment::redemption(&investor, investment_id)?;
```

Comment:
This amount might contain be below the actual `amount_unprocessed_investment ` amount. The reason being is, that a user needs to collect before this amount reflects the actual `amount_unprocessed_investment`. 

Is this okay?

Otherwise we need to
* execute `collect` in `rame_support::storage::with_storage_layer` if we want the state change to be reverted that collect is introducing
* execute `collect` 

Note: In both cases we must check that we are `FullyCollected`.

Reply:
Say a user needs to call `collect` more than once to collect everything. I would hope that in between both calls the `T::Investment::redemption` (and respectively `T::Investment::investment`) amounts do not change?

Reply:
They can only if a user calls `collect` or if a user is already fully collecte if they change their order.

Reply:
[2beb562](https://github.com/centrifuge/centrifuge-chain/pull/1473/commits/2beb562d0d41172440b01ff7203d76ce1f1b7fb7)